### PR TITLE
IPv6 support

### PR DESCRIPTION
**Pull request by altergui**
this is a work-in-progress exploring how would dualstack support mean in ipfsSync.go
```
2021-10-11T14:43:34+02:00	INFO	ipfssync/ipfssync.go:412	my multiaddress ipv4: /ip4/95.xxx.xxx.2/tcp/4001/p2p/QmTj3Zrrk56jYYpwtdRWzKFFHCZM1HdujTVRRTrm6UfRn9
2021-10-11T14:43:34+02:00	INFO	ipfssync/ipfssync.go:415	my multiaddress ipv6: /ip6/2a02:xxxx:fede::feca/tcp/4001/p2p/QmTj3Zrrk56jYYpwtdRWzKFFHCZM1HdujTVRRTrm6UfRn9
2021-10-11T14:43:34+02:00	DEBUG	ipfssync/ipfssync.go:223	broadcasting message HELLO {Address:QmRH37orTHrt3e8xcE3rFi7jdHBw3tJf1AThgU9GyrmwsD Hash: MA:/ip4/95.xxx.xxx.2/tcp/4001/p2p/QmTj3Zrrk56jYYpwtdRWzKFFHCZM1HdujTVRRTrm6UfRn9 PL:[] Ts:1633956214}
2021-10-11T14:43:34+02:00	DEBUG	subpubtransport/subpub.go:88	sending 135 bytes to broadcast channel
2021-10-11T14:43:34+02:00	DEBUG	ipfssync/ipfssync.go:223	broadcasting message HELLO {Address:QmRH37orTHrt3e8xcE3rFi7jdHBw3tJf1AThgU9GyrmwsD Hash: MA:/ip6/2a02:xxxx:fede::feca/tcp/4001/p2p/QmTj3Zrrk56jYYpwtdRWzKFFHCZM1HdujTVRRTrm6UfRn9 PL:[] Ts:1633956214}
2021-10-11T14:43:34+02:00	DEBUG	subpubtransport/subpub.go:88	sending 152 bytes to broadcast channel
2021-10-11T14:43:34+02:00	DEBUG	subpub/discovery.go:43	found peer: QmatPDpxw8ovckRZ7vJkp7aPXuv5xxFeBXaCTtL78QtG6r
2021-10-11T14:43:34+02:00	INFO	subpub/stream.go:40	connected to peer QmatPDpxw8ovckRZ7vJkp7aPXuv5xxFeBXaCTtL78QtG6r: /ip6/2a02:xxxx::1/tcp/4171
```

### CODE REVIEW

File: ipfssync/ipfssync.go
Comment (mvdan) [inline ipfssync/ipfssync.go:410]: do we want to allow having both be non-nil? your code does that, just want to double check.

File: util/net.go
Comment (mvdan) [inline util/net.go:19]: Hmm, where is this defined? Shouldn't the `ipversion` be passed to `ExternalIP`?

**Comment by altergui** [comment]
hey, thanks a lot for the review!! i'm just slowly starting to get a grasp on all this, so getting some feedback makes me feel a little bit less lost :) and is much appreciated

File: ipfssync/ipfssync.go
Comment (altergui) [reply ipfssync/ipfssync.go:410]: yes, if both ipv4 and ipv6 are discovered, it will send two HELLOs, one with /ip4/ and one with /ip6/, but pointing to the same nodeID (last part of the multiaddr), which in essence means "you can reach me over these two addresses". 

it should only fail if *no* addresses are discovered (so both err4 and err6 are non-nil)
if either ipv4 or ipv6 is available, that's OK.

File: util/net.go
Comment (altergui) [reply util/net.go:19]: i'm using @jc21 syntax since i `replace`d go-external-ip in go.mod
https://github.com/vocdoni/vocdoni-node/pull/313/files/c871e212feb89c1cf3d26795850c771ab63e50ab#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R12

Comment (mvdan) [reply util/net.go:19]: Ah I see. Given that upstream will likely review/merge that PR in the next few days, I'd say we should wait until that happens to merge this, because we should hopefully be able to get rid of our fork :)

**Comment by mvdan** [comment]
Looks like upstream did a release with this feature :) https://github.com/GlenDC/go-external-ip/releases/tag/v0.1.0

**Comment by altergui** [comment]
rebased, squashed, pointed go-external-ip to upstream, and retested, still good :)

File: ipfssync/ipfssync.go
Comment (mvdan) [inline ipfssync/ipfssync.go:63]: re-run gofmt?

File: ipfssync/ipfssync.go
Comment (mvdan) [inline ipfssync/ipfssync.go]: Hmm, would be nice if we defined these constants in a single place, but I guess we don't have a good place in this module. Perhaps if upstream exposed the constants it would be better.

**Review by mvdan — APPROVED** [review]
SGTM!

File: ipfssync/ipfssync.go
Comment (altergui) [reply ipfssync/ipfssync.go:63]: right

**Comment by altergui** [comment]
gofmt'd

**Comment by mvdan** [comment]
SGTM. We should probably squash-merge, unless you want to squash your commits yourself.

**Comment by altergui** [comment]
sounds good, squashed

**Review by mvdan — APPROVED** [review]
LGTM! Up to you if you want a review from @p4u as well. If not, we can merge on green.

**Comment by emmdim** [comment]
@mvdan @altergui @jordipainan  any concrete reason the compose-test is fail or should we simply just re-run the test?

**Comment by p4u** [comment]
I am re-running the compose test, lets see.

@altergui you can run the compose test in your local machine like this too:
```
cd dockerfiles/testsuite
bash start_test.sh
```

EDIT: you will need support for docker buildKit ( `{ "features": { "buildkit": true } }` >> /etc/docker/daemon.json )

**Comment by altergui** [comment]
i did introduce a small (but critical) regression in this PR:
setting ListenAddr to `/ip6/::/tcp/%d` does NOT make it dualstack but ipv6-only!
for listening on ipv4 as well, `/ip4/0.0.0.0/tcp/%d` needs to be added as well. 

Sorry about that, I squashed this bug in https://github.com/vocdoni/vocdoni-node/pull/363 (specifically in [this line](https://github.com/vocdoni/vocdoni-node/commit/936aa8dd9182ebb9584d0b079cf1fa7201a0bf5f#diff-e28cfee6e67f3ae610cb00a944329f65d646e1b36c5a356a18b8c25df8cc8ca5R129))

**Pull request merged**
